### PR TITLE
Fourier series representation of chromatic noise (`CMWaveX`)

### PR DESCRIPTION
**Pull request by abhisrkckl**
See #1661.

Like `WaveX` & `DMWaveX`

#1797 needs to be merged first. I have merged that PR into this one to implement some validation steps that depend on both `CMWaveX` and `PLChromNoise`.

### CODE REVIEW

**Comment by abhisrkckl** [comment]
Here is a comparison between PINT and ENTERPRISE analyses for chromatic noise to ensure the conventions are consistent. This uses a simulated dataset.

**Harmonic number selection**
image

**Parameter estimation**
image
Red point is PINT estimate
Blue lines are injected values
Black corner plot is the ENTERPRISE posterior.

**Chromatic noise spectrum**
image

File: src/pint/models/binary_ddk.py [hunk 1]
```diff
@@ -49,14 +49,19 @@ class BinaryDDK(BinaryDD):
     of the system from Earth, the finite size of the system, and the
     interaction of these with the proper motion.
 
-    From Kopeikin (1995) this includes :math:`\Delta_{\pi M}` (Equation 17), the mixed annual-orbital parallax term, which changes :math:`a_1` and :math:`\omega`
-    (:meth:`~pint.models.stand_alone_psr_binaries.DDK_model.DDKmodel.delta_a1_parallax` and :meth:`~pint.models.stand_alone_psr_binaries.DDK_model.DDKmodel.delta_omega_parallax`).
+    From Kopeikin (1995) this includes :math:`\\Delta_{\\pi M}` (Equation 17),
```
Comment (dlakaplan) [inline src/pint/models/binary_ddk.py:52]: Maybe get rid of the double \\ and convert to `r""""""`?

File: tests/test_cmwavex.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        F1     -4.330899370129D-14  1  2.149749089617D-22
+        PEPOCH        55321.000000
+        DM               71.016633
+        CM                     0.1
+        TNCHROMIDX               4
+        UNITS                  TDB
+    """
+    m = get_model(StringIO(par))
+
+    with pytest.raises(ValueError):
```
Comment (dlakaplan) [inline tests/test_cmwavex.py:34]: should this be a different test than the other pieces?

File: tests/test_cmwavex.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    t = make_fake_toas_uniform(54000, 56000, 200, m, add_noise=True)
+
+    ftr = Fitter.auto(t, m)
+    ftr.fit_toas()
+
+    assert ftr.resids.reduced_chi2 < 2
+
+
+def test_cmwavex_badpar():
```
Comment (dlakaplan) [inline tests/test_cmwavex.py:53]: should these tests be somehow separated into different pieces?

File: tests/test_cmwavex.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    assert m.num_cmwavex_freqs == len(idxs) + 3
+
+    m.components["CMWaveX"].add_cmwavex_components(
+        np.array([3]) / u.day,
+        cmwxsins=np.array([0]) * cmu,
+        cmwxcoses=np.array([0]) * cmu,
+    )
+    assert m.num_cmwavex_freqs == len(idxs) + 4
+
+    with pytest.raises(ValueError):
```
Comment (dlakaplan) [inline tests/test_cmwavex.py:161]: I think all of these statements that raise exceptions should be in separate tests?  So the main test should have statements that succeed, and each failure/exception should be separate?

**Comment by dlakaplan** [comment]
It looks like almost all of the changes to `timing_model` were general code updates not related to this PR.  Is that right?   I didn't examine them all in detail, but I'm guessing if all of the tests pass then they are probably OK.

**Comment by dlakaplan** [comment]
> Here is a comparison between PINT and ENTERPRISE analyses for chromatic noise to ensure the conventions are consistent. This uses a simulated dataset.
> 
> This is similar to the simulation studies presented in https://arxiv.org/abs/2405.01977
> 
> **Harmonic number selection** image
> 
> **Parameter estimation** image Red point is PINT estimate Blue lines are injected values Black corner plot is the ENTERPRISE posterior.
> 
> **Chromatic noise spectrum** image

This is nice.  Since it didn't make it into the paper, and this PR will be hard to find once it is merged, will these results be preserved somehow?  Maybe a notebook (but not one that is automatically run unless it is fast)?

File: src/pint/models/binary_ddk.py (hunk 1 above)
Comment (abhisrkckl) [reply src/pint/models/binary_ddk.py]: changed.

File: tests/test_cmwavex.py (hunk 4 above)
Comment (abhisrkckl) [reply tests/test_cmwavex.py:161]: Moved to separate tests

File: tests/test_cmwavex.py (hunk 2 above)
Comment (abhisrkckl) [reply tests/test_cmwavex.py]: split into a different test.

File: tests/test_cmwavex.py (hunk 3 above)
Comment (abhisrkckl) [reply tests/test_cmwavex.py:62]: Split the fitting test into a separate function.

**Comment by abhisrkckl** [comment]
> It looks like almost all of the changes to timing_model were general code updates not related to this PR. Is that right? I didn't examine them all in detail, but I'm guessing if all of the tests pass then they are probably OK.

Only the changes to `TimingModel.validate_component_types()` method are related to this PR. Everything else is refactorings suggested by sourcery.

**Comment by dlakaplan** [comment]
OK, I think the code looks fine here.  Any thoughts on the docs?

**Comment by abhisrkckl** [comment]
> OK, I think the code looks fine here. Any thoughts on the docs?

Example added to notebook.
https://nanograv-pint--1802.org.readthedocs.build/en/1802/examples/rednoise-fit-example.html#Chromatic-noise-fitting

**Comment by dlakaplan** [comment]
So is that notebook one that is built & run as part of CI?   If so, any estimate of the additional run-time?

**Comment by abhisrkckl** [comment]
~10 s on my laptop.

**Comment by dlakaplan** [comment]
OK, good.  I can merge when ready then.

**Comment by abhisrkckl** [comment]
Ok... please do.

**Pull request merged**